Make "plugin list" list plugins by context

marckhouzam opened the pull request:
### What this PR does / why we need it

This PR is based on top of the one commit of #97, but only the top commit should be of interest for this PR.

This commit changes "plugin list" to show plugins by context.
This is the same format that was used before the Central Repo feature.
The differences with the previous version (pre-central-repo) are:
1. only installed standalone plugins are shown
2. missing context plugins are also listed
    
### Which issue(s) this PR fixes
<!--
     Usage: Fixes #<issue number>.

     Unless the PR is for a trivial change (e.g. fixing a typo), consider opening an issue first
     (and reference it here) so that the problem the PR addresses can be discussed independently of
     the solutions proposed by this PR.
-->
Fixes #98 

### Describe testing done for PR

```
$ rm ~/.config/tanzu/config*
$ tz plugin clean
[ok] successfully cleaned up all plugins
$ tz config set features.global.central-repository true
$ export TANZU_CLI_PRE_RELEASE_REPO_IMAGE=localhost:9876/tanzu-cli/plugins/central:small
$ make start-test-central-repo
[...]

$ tz context create --name tmc --endpoint unstable.tmc-dev.cloud.vmware.com --staging
[i] API token env var is set

[ok] successfully created a TMC context
[i] Checking for required plugins...
[i] Installing plugin 'iam:v0.0.1' with target 'mission-control'
[i] Installing plugin 'continuousdelivery:v0.0.1' with target 'mission-control'
[i] Installing plugin 'inspection:v0.0.1' with target 'mission-control'
[...]
[i] Successfully installed all required plugins

$ tz plugin list
Installed Standalone Plugins
  NAME  DESCRIPTION  TARGET  VERSION  STATUS

Plugins from Context:  tmc
  NAME                DESCRIPTION                                                     TARGET           VERSION  STATUS
  account             Account for tmc resources                                       mission-control  v0.0.1   installed
  apply               Create/Update a resource with a resource file                   mission-control  v0.0.1   installed
  audit               Run an audit request on an org                                  mission-control  v0.0.1   installed
  [...]
  tanzupackage        Tanzupackage for tmc resources                                  mission-control  v0.0.1   installed
  workspace           A group of Kubernetes namespaces                                mission-control  v0.0.1   installed

$ tz context create --name k3d --kubecontext k3d-k3s-default --kubeconfig /Users/kmarc/.config/k3d/k3s-default/kubeconfig.yaml
[ok] successfully created a kubernetes context using the kubeconfig /Users/kmarc/.config/k3d/k3s-default/kubeconfig.yaml
[i] Checking for required plugins...
[i] Installing plugin 'cluster:v9.9.9' with target 'kubernetes'
[i] Installing plugin 'feature:v9.9.9' with target 'kubernetes'
[i] Installing plugin 'kubernetes-release:v0.0.1' with target 'kubernetes'
[i] Successfully installed all required plugins
$
$ tz plugin list
Installed Standalone Plugins
  NAME  DESCRIPTION  TARGET  VERSION  STATUS

Plugins from Context:  k3d
  NAME                DESCRIPTION                           TARGET      VERSION  STATUS
  cluster             Kubernetes cluster operations         kubernetes  v9.9.9   installed
  feature             Operate on features and featuregates  kubernetes  v9.9.9   installed
  kubernetes-release  Kubernetes release operations         kubernetes  v0.0.1   installed

Plugins from Context:  tmc
  NAME                DESCRIPTION                                                     TARGET           VERSION  STATUS
  account             Account for tmc resources                                       mission-control  v0.0.1   installed
  apply               Create/Update a resource with a resource file                   mission-control  v0.0.1   installed
  audit               Run an audit request on an org                                  mission-control  v0.0.1   installed
[...]
  tanzupackage        Tanzupackage for tmc resources                                  mission-control  v0.0.1   installed
  workspace           A group of Kubernetes namespaces                                mission-control  v0.0.1   installed

$ tz plugin install isolated-cluster
[i] Installing plugin 'isolated-cluster:v9.9.9' with target 'global'
[ok] successfully installed 'isolated-cluster' plugin
$ tz plugin list
Installed Standalone Plugins
  NAME              DESCRIPTION                     TARGET  VERSION  STATUS
  isolated-cluster  isolated-cluster functionality  global  v9.9.9   installed

Plugins from Context:  k3d
  NAME                DESCRIPTION                           TARGET      VERSION  STATUS
  cluster             Kubernetes cluster operations         kubernetes  v9.9.9   installed
  feature             Operate on features and featuregates  kubernetes  v9.9.9   installed
  kubernetes-release  Kubernetes release operations         kubernetes  v0.0.1   installed

Plugins from Context:  tmc
  NAME                DESCRIPTION                                                     TARGET           VERSION  STATUS
  account             Account for tmc resources                                       mission-control  v0.0.1   installed
  apply               Create/Update a resource with a resource file                   mission-control  v0.0.1   installed
  audit               Run an audit request on an org                                  mission-control  v0.0.1   installed
[...]
  tanzupackage        Tanzupackage for tmc resources                                  mission-control  v0.0.1   installed
  workspace           A group of Kubernetes namespaces                                mission-control  v0.0.1   installed

$ tz plugin delete workspace
Deleting Plugin 'workspace'. Are you sure? [y/N]: y
[ok] successfully deleted plugin 'workspace'
$ tz plugin delete feature
Deleting Plugin 'feature'. Are you sure? [y/N]: y
[ok] successfully deleted plugin 'feature'
$ tz plugin list
Installed Standalone Plugins
  NAME              DESCRIPTION                     TARGET  VERSION  STATUS
  isolated-cluster  isolated-cluster functionality  global  v9.9.9   installed

Plugins from Context:  k3d
  NAME                DESCRIPTION                    TARGET      VERSION  STATUS
  cluster             Kubernetes cluster operations  kubernetes  v9.9.9   installed
  kubernetes-release  Kubernetes release operations  kubernetes  v0.0.1   installed

Plugins from Context:  tmc
  NAME                DESCRIPTION                                                     TARGET           VERSION  STATUS
  account             Account for tmc resources                                       mission-control  v0.0.1   installed
  apply               Create/Update a resource with a resource file                   mission-control  v0.0.1   installed
  audit               Run an audit request on an org                                  mission-control  v0.0.1   installed
[...]
  secret              secret for tmc resources                                        mission-control  v0.0.1   installed
  tanzupackage        Tanzupackage for tmc resources                                  mission-control  v0.0.1   installed

Missing Plugins from Context:  k3d
  NAME     DESCRIPTION                           TARGET      VERSION  STATUS
  feature  Operate on features and featuregates  kubernetes  v9.9.9   not installed

Missing Plugins from Context:  tmc
  NAME       DESCRIPTION                       TARGET           VERSION  STATUS
  workspace  A group of Kubernetes namespaces  mission-control  v0.0.1   not installed

[!] As shown above, some recommended plugins have not been installed. To install them please run 'tanzu plugin sync'.

$ tz plugin list -o yaml
- context: ""
  description: isolated-cluster functionality
  name: isolated-cluster
  status: installed
  target: global
  version: v9.9.9
- context: k3d
  description: Kubernetes cluster operations
  name: cluster
  status: installed
  target: kubernetes
  version: v9.9.9
- context: k3d
  description: Kubernetes release operations
  name: kubernetes-release
  status: installed
  target: kubernetes
  version: v0.0.1
- context: tmc
  description: Account for tmc resources
  name: account
  status: installed
  target: mission-control
  version: v0.0.1
- context: tmc
  description: Create/Update a resource with a resource file
  name: apply
  status: installed
  target: mission-control
  version: v0.0.1
[...]
- context: tmc
  description: Tanzupackage for tmc resources
  name: tanzupackage
  status: installed
  target: mission-control
  version: v0.0.1
- context: k3d
  description: Operate on features and featuregates
  name: feature
  status: not installed
  target: kubernetes
  version: v9.9.9
- context: tmc
  description: A group of Kubernetes namespaces
  name: workspace
  status: not installed
  target: mission-control
  version: v0.0.1
```

#### Special notes for your reviewer

1. The title of the installed plugin section is "Installed Standalone Plugins" (pre central-repo it was "Standalone Plugins")
2. Should we get rid of the "status" column?  I feel it makes it more obvious to users when plugins are missing, if they miss the section title "Missing Plugins from Context:", but that column is 100% redundant.
3. Installing the same plugin both as standalone (manually) and context-scope (through a context) does not work well yet. I'll post a separate PR for that.

anujc25 commented:
Instead of showing different sections for `Missing Plugins from Context` should we show those plugins under the same `Plugins from Context:` section with `STATUS` being `not installed`?

anujc25 commented:
> The title of the installed plugin section is "Installed Standalone Plugins" (pre central-repo it was "Standalone Plugins")
Should we get rid of the "status" column? I feel it makes it more obvious to users when plugins are missing, if they miss the section title "Missing Plugins from Context:", but that column is 100% redundant.

My suggestion is we combine the installed and missing plugin sections and keep the status as it is. Here is the sample:
```
Standalone Plugins
  NAME              DESCRIPTION                     TARGET  VERSION  STATUS
  isolated-cluster  isolated-cluster functionality  global  v9.9.9   installed

Plugins from Context:  k3d
  NAME                DESCRIPTION                    TARGET      VERSION  STATUS
  cluster             Kubernetes cluster operations  kubernetes  v9.9.9   installed
  kubernetes-release  Kubernetes release operations  kubernetes  v0.0.1   installed
  feature             Operate on features            kubernetes  v9.9.9   not installed

Plugins from Context:  tmc
  NAME                DESCRIPTION                                  TARGET           VERSION  STATUS
  account             Account for tmc resources                   mission-control  v0.0.1   installed
  apply               Create/Update a resource with a resource    mission-control  v0.0.1   installed
  audit               Run an audit request on an org              mission-control  v0.0.1   installed
[...]
  secret              secret for tmc resources                    mission-control  v0.0.1   installed
  tanzupackage        Tanzupackage for tmc resources              mission-control  v0.0.1   installed
  workspace           A group of Kubernetes namespaces            mission-control  v0.0.1   not installed
```

marckhouzam commented:
> My suggestion is we combine the installed and missing plugin sections and keep the status as it is.

Good idea!  I will update the PR.  Thanks!

marckhouzam commented:
> > My suggestion is we combine the installed and missing plugin sections and keep the status as it is.
> 
> Good idea! I will update the PR. Thanks!

I have updated the PR and the PR description.

I want to point out a possible confusing aspect of this change.
For the "Installed Standalone Plugins" section, we only show _installed_ plugins, but for the context sections we show both installed and not installed plugins. 
I wonder if a user may assume that the "Installed Standalone Plugins" section also shows missing plugins (users often don't pay attention to the title in much detail).  To imply that this section does not report different statuses, I removed the "status" column, just from that section.

Please let me know if you have a different opinion about this.

marckhouzam commented:
I got some feedback that it would be better to be consistent with the columns so I have re-introduced the "status" column for installed plugins.

vuil commented on pkg/command/plugin.go:
```diff
@@ -545,93 +556,98 @@ func displayPluginListOutputSplitViewContext(availablePlugins []discovery.Discov
-			installedPlugins[index].Description,
-			string(installedPlugins[index].Target),
-			installedPlugins[index].Version,
+	_, _ = cyanBold.Println("Installed Standalone Plugins")
```
Did you mean to keep the "Installed "?

marckhouzam replied:
Will remove.

vuil reviewed (approved):
lgtm, thanks! 
just have a few nits.

anujc25 commented on pkg/pluginmanager/manager.go:
```diff
@@ -545,48 +545,103 @@ func InitializePlugin(plugin *cli.PluginInfo) error {
+// InstallPluginFromContext installs a plugin by name, version and target as a context-scope plugin.
+func InstallPluginFromContext(pluginName, version string, target configtypes.Target, contextName string) error {
+	if contextName == "" {
+		log.V(4).Warning("Missing context name for a context-scope plugin: %s/%s/%s", pluginName, version, string(target))
```
Currently, we do not have a way to select the log level when invoking the command. So, this log should be printed at normal log level without `V(4)`.

marckhouzam replied:
If we hit this case it probably means we have done a coding error where we called `InstallPluginFromContext()`.
So as Vui says, there is probably nothing a user can do, which is why I didn't want it to show all the time.  But I did want to have some way to help troubleshoot it...

I will remove the `V(4)` since we don't have a way to configure the logging level yet.

anujc25 reviewed (commented):
Overall LGTM.
